### PR TITLE
Markets emails

### PR DESCRIPTION
**Pull request by IanPhilips**
I have to merge this with main to see how it looks with real markets (only sending the email to myself) but if that looks good will just have to change the interval and sub out the query for me for the query for all users.

The function gets all private users and filters for those who haven't unsubscribed from weekly trending emails (unsubscribe link at bottom), and gets the top 50 trending markets by unique individual bettor ids. For each user it filters the top 50 markets for ones they haven't bet in already and sends them 3 random ones. I was thinking about tracking the ones we've sent them (just a string array of slugs on the PrivateUser model) but Stephen thinks this isn't necessary so I removed it. 

I had to move a few different bits of code from web/lib to common so that I could build the og-image on the backend and send it with the email.

Screen Shot 2022-08-17 at 11 38 18 AM

### CODE REVIEW

File: functions/src/emails.ts [hunk 1]
```diff
@@ -390,3 +391,53 @@ export const sendNewAnswerEmail = async (
...
+export const sendThreeContractsEmail = async (
+  privateUser: PrivateUser,
+  contractsToSend: Contract[]
+) => {
+  const emailType = 'weekly-trending'
+  const unsubscribeUrl = `${UNSUBSCRIBE_ENDPOINT}?id=${privateUser.id}&type=${emailType}`
+  if (!privateUser || !privateUser.email) return
+  await sendTemplateEmail(
+    privateUser.email,
+    contractsToSend[0].question + ' and 2 more questions for you.',
```
Comment (mantikoros) [inline functions/src/emails.ts:404]: Not sure about this as an email subject. Something like "Interesting markets this week" would be better, I think.

File: functions/src/emails.ts [hunk 2]
```diff
@@ -390,3 +391,53 @@ export const sendNewAnswerEmail = async (
...
   )
 }
+
+export const sendThreeContractsEmail = async (
+  privateUser: PrivateUser,
+  contractsToSend: Contract[]
+) => {
+  const emailType = 'weekly-trending'
+  const unsubscribeUrl = `${UNSUBSCRIBE_ENDPOINT}?id=${privateUser.id}&type=${emailType}`
+  if (!privateUser || !privateUser.email) return
```
Comment (mantikoros) [inline functions/src/emails.ts:401]: You should check if the privateUser is unsubscribed here.

File: functions/src/weekly-markets-emails.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import { Contract } from '../../common/contract'
+import { getPrivateUser, getValues, isProd, log } from './utils'
+import { filterDefined } from '../../common/util/array'
+import { sendThreeContractsEmail } from './emails'
+import { createRNG, shuffle } from '../../common/util/random'
+import { DAY_MS } from '../../common/util/time'
+
+export const weeklyMarketsEmails = functions
+  .runWith({ secrets: ['MAILGUN_KEY'] })
+  .pubsub.schedule('every 1 minutes')
```
Comment (mantikoros) [inline functions/src/weekly-markets-emails.ts:13]: This is just debugging code, I assume. When do you plan to send this email? I vote at a consistent time like Monday at 12 pm ET.

File: functions/src/weekly-markets-emails.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+const firestore = admin.firestore()
+
+async function getTrendingContracts() {
+  return await getValues<Contract>(
+    firestore
+      .collection('contracts')
+      .where('isResolved', '==', false)
+      .where('closeTime', '>', Date.now() + DAY_MS)
+      .where('visibility', '==', 'public')
```
Comment (mantikoros) [inline functions/src/weekly-markets-emails.ts:26]: Do we even use the visibility flag?

**Review by mantikoros — COMMENTED** [review]
Very nice! Looks like this basically ready to go.

A few additional comments:
- The "see more" text in the email template is awkwardly positioned. I would just remove it entirely.
- Please create a file for the email template at functions/src/email-templates

**Comment by IanPhilips** [comment]
The see more text is only weird for markets without descriptions. Most have descriptions and will look like the last market. Do you think the last market's see more button is awkward?

File: functions/src/emails.ts (hunk 1 above)
Comment (IanPhilips) [reply functions/src/emails.ts:404]: If they're truly interesting then one of the titles should get their attention, right?

File: functions/src/emails.ts (hunk 2 above)
Comment (IanPhilips) [reply functions/src/emails.ts:401]: Well I do check that in the `weekly-markets-email` function which is what that flag is for. You can use this `sendThreeContractsEmail` function and check for your own flag.

Comment (IanPhilips) [reply functions/src/emails.ts:401]: I suppose you'll use a different template though so I'll add it again here

File: functions/src/weekly-markets-emails.ts (hunk 4 above)
Comment (IanPhilips) [reply functions/src/weekly-markets-emails.ts:26]: No time like the present! It is set on markets currently and we've been talking about using it.

File: functions/src/weekly-markets-emails.ts (hunk 3 above)
Comment (IanPhilips) [reply functions/src/weekly-markets-emails.ts:13]: yup, just for debugging. I'll set it to that time after I merge it into main and uncomment the all users bit.

File: functions/src/emails.ts (hunk 1 above)
Comment (jahooma) [reply functions/src/emails.ts:404]: I like Ian's idea to include the question.

What do you think about sending more markets? Like 6?

Comment (IanPhilips) [reply functions/src/emails.ts:404]: hmmm!! yeah we might as well, for all we know these could be the only markets they see all week

**Comment by jahooma** [comment]
> The see more text is only weird for markets without descriptions. Most have descriptions and will look like the last market. Do you think the last market's see more button is awkward?

Maybe the description / see more paragraph should be aligned directly under the image (same width)

**Comment by IanPhilips** [comment]
Hm I can't figure out how to download the html from the preview. I used the template builder and when I preview the content It's in an iframe but can't figure out how to download the iframe contents. Not going to block on that as they're just backups it seems like - they're stored on mailgun already.

**Comment by IanPhilips** [comment]
I moved it to be centered after the description text but it sort of feels like it might take me to the next question so unsure of how much better it is.
Screen Shot 2022-08-17 at 4 58 22 PM

**Comment by mantikoros** [comment]
Yeah, I think the market description is not adding anything. I'd remove it.

I'll figure out how to extract the html from the email. Feel free to merge this in and start running now.

**Comment by mantikoros** [comment]
I created an html email with my design:
Screen Shot 2022-08-19 at 11 14 18 AM

**Pull request merged**